src/core/thread/network_data_publisher.cpp

```diff
@@ -0,0 +1,492 @@
+ *
+ */
+
+#include "network_data_publisher.hpp"
+
+#if OPENTHREAD_CONFIG_NETDATA_PUBLISHER_ENABLE
+
+#include "common/code_utils.hpp"
+#include "common/instance.hpp"
+#include "common/locator_getters.hpp"
+#include "common/random.hpp"
+#include "thread/network_data_local.hpp"
+#include "thread/network_data_service.hpp"
+
+namespace ot {
+namespace NetworkData {
+
+Publisher::Publisher(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mState(kNoEntry)
+    , mTimer(aInstance, Publisher::HandleTimer)
+    , mCallback(nullptr)
+    , mContext(nullptr)
+{
+}
+
+void Publisher::PublishDnsSrpServiceAnycast(uint8_t aSequenceNumber)
+{
+    otLogInfoNetData("Publisher: Publishing DNS/SRP service anycast (seq-num:%d)", aSequenceNumber);
+    Publish(kTypeAnycast, aSequenceNumber, nullptr);
+}
+
+void Publisher::PublishDnsSrpServiceUnicast(const Ip6::Address &aAddress, uint16_t aPort)
+{
+    otLogInfoNetData("Publisher: Publishing DNS/SRP service unicast (%s, port:%d)", aAddress.ToString().AsCString(),
+                     aPort);
+    Publish(kTypeUnicast, aPort, &aAddress);
+}
+
+void Publisher::PublishDnsSrpServiceUnicast(uint16_t aPort)
+{
+    otLogInfoNetData("Publisher: Publishing DNS/SRP service unicast (ml-eid, port:%d)", aPort);
+    Publish(kTypeUincastMeshLocalEid, aPort, &Get<Mle::Mle>().GetMeshLocal64());
+}
+
+void Publisher::SetDnsSrpServiceCallback(DnsSrpServiceCallback aCallback, void *aContext)
+{
+    mCallback = aCallback;
+    mContext  = aContext;
+}
+
+void Publisher::UnpublishDnsSrpService(void)
+{
+    otLogInfoNetData("Publisher: Unpublishing DNS/SRP service");
+
+    Remove(/* aNextState */ kNoEntry);
+}
+
+void Publisher::Publish(Type aType, uint16_t aPortOrSeqNumber, const Ip6::Address *aAddress)
+{
+    if (mState != kNoEntry)
+    {
+        if ((mType == aType) && (aPortOrSeqNumber == mPortOrSeqNumber) &&
+            ((aAddress == nullptr) || (*aAddress == mAddress)))
+        {
+            otLogInfoNetData("Publisher: Same service is already being published");
+            ExitNow();
+        }
+
+        Remove(/* aNextState */ kNoEntry);
+    }
+
+    mType            = aType;
+    mPortOrSeqNumber = aPortOrSeqNumber;
+
+    if (aAddress != nullptr)
+    {
+        mAddress = *aAddress;
+    }
+
+    SetState(kToAdd);
+    Process();
+
+exit:
+    return;
+}
+
+void Publisher::SetState(State aState)
+{
+    State oldState = mState;
+
+    VerifyOrExit(mState != aState);
+
+    otLogInfoNetData("Publisher: State: %s -> %s", StateToString(mState), StateToString(aState));
+    mState = aState;
+
+    // Invoke callback to notify the change to the Network Data, when
+    // the entry is added (on transition to new state `kAdded`), or
+    // when it is removed (on transition from state `kAdded`).
+
+    VerifyOrExit((mState == kAdded) || (oldState == kAdded));
+
+#if OPENTHREAD_CONFIG_SRP_SERVER_ENABLE
+    Get<Srp::Server>().HandleNetDataPublisherEntryChange(mState == kAdded);
```

Comment:
Seems we are turning off SRP server when `mState` is not `kAdded`. 

However, if `mState` is changing from `kAdded` to `kRemoving`, it's not actually removed from `Network Data`, yet. 
And it's also possible for Publisher to revert the state from `kRemoving` to `kAdded` to cancel the removal. 

Should we keep the SRP server running in such case since the Service TLV is still in `Network Data` ? Clients may still send registrations.

Reply:
Good point! We can stop the server too early.

In new push, I changed how and when `Publisher` notifies (emits events  and invokes callback). I kept it as a separate commit.